### PR TITLE
Stop Skipping Asset Precompilation on Drone

### PR DESCRIPTION
**Pull request by Hamms**
For better consistency with other environments.

I added some temporary benchmarking in https://github.com/code-dot-org/code-dot-org/pull/46996/commits/84373e5c16ff8fe4cd0f36a55e8a19f139fbf00a to see how much, and this does add an average of about 3.5 minutes to the drone builds. I think that's acceptably low, but am open to feedback.

## Links

See https://github.com/code-dot-org/code-dot-org/pull/31153, which did some work in this space and referenced unkown issues with these settings on drone. I'm not able to reproduce the issues referenced; I expect that they vanished just as mysteriously as they appeared.

## PR Checklist:

- [ ] Tests provide adequate coverage
- [ ] Privacy and Security impacts have been assessed
- [ ] Code is well-commented
- [ ] New features are translatable or updates will not break translations
- [ ] Relevant documentation has been added or updated
- [ ] User impact is well-understood and desirable
- [ ] Pull Request is labeled appropriately
- [ ] Follow-up work items (including potential tech debt) are tracked and linked

### CODE REVIEW

**Comment by davidsbailey** [comment]
> For better consistency with other environments.
> 
> I added some temporary benchmarking in [84373e5](https://github.com/code-dot-org/code-dot-org/commit/84373e5c16ff8fe4cd0f36a55e8a19f139fbf00a) to see how much, and this does add an average of about 3.5 minutes to the drone builds. I think that's acceptably low, but am open to feedback.

Thank you for taking the time to profile this! While it is not a total dealbreaker, 3.5 minutes is a tangible cost, especially as we have a long-term goal of reducing our CI build/test time down to something ambitiously low. I'm having trouble weighing this tangible time cost against the less-tangible benefit of better consistency. did you have any more specific ideas around the potential risks of not doing this (or other potential benefits)?

my assumption is that this change primarily affects whether we do `yarn build` or `yarn build:dist` before running apps tests, because we already set `optimize_webpack_assets` to `false` in rails tests ([here](https://github.com/code-dot-org/code-dot-org/blob/92956268b82a63a3f6e17e178052497365fef687/dashboard/test/test_helper.rb#L89)), and (AFAIK?) we do not compile or digest rails assets during rails unit tests. If that does not sound right please let me know.

I don't want this to get stuck because one person leans no -- if you do still want to proceed with this, my other requests would be (1) to get another opinion from someone on the developer tools working group to see if this matters to them, and (2) to consider leaving a comment behind indicating that the reasons we are taking a hit on build time (so that the option of undoing this change will be discoverable in the future).

> See #31153, which did some work in this space and referenced unkown issues with these settings on drone. I'm not able to reproduce the issues referenced; I expect that they vanished just as mysteriously as they appeared.

Agreed on not worrying about those issues, because it sounds like they were related to changing the same params, but in the opposite direction and on the other drone container.

**Comment by Hamms** [comment]
The underlying motivation here is that we want to change a Rails configuration setting which raises errors on unknown assets, in preparation for updating to Rails 6.1 (https://github.com/code-dot-org/code-dot-org/pull/46937). This setting dramatically increases the difference in behavior between environments with and without compiled assets.

If we _don't_ compile assets on Drone, then we'll need to keep that setting disabled on Drone and so will increase the risk of legitimate test failures making it to the DTT.

I should also clarify that this adds 3.5 minutes specifically to the unit test container, which right now takes about 40 minutes less time to complete than the UI test container.

**Comment by davidsbailey** [comment]
> The underlying motivation here is that we want to change a Rails configuration setting which raises errors on unknown assets, in preparation for updating to Rails 6.1 (#46937). This setting dramatically increases the difference in behavior between environments with and without compiled assets.
> 
> If we _don't_ compile assets on Drone, then we'll need to keep that setting disabled on Drone and so will increase the risk of legitimate test failures making it to the DTT.

Thanks for this context. If you are sure you need it for the rails upgrade, then this is fine with me. I would just like to add that there are a lot of things being lumped together here, not by you but by the existing code:
* https://github.com/code-dot-org/code-dot-org/blob/cd77047c9cdd479881816b64d15b2715bfd84edb/dashboard/config/application.rb#L125-L129
* https://github.com/code-dot-org/code-dot-org/blob/cbad9b3b9c8b1531dabf8aeb792d7049b2eaf744/lib/rake/build.rake#L97-L101
* https://github.com/code-dot-org/code-dot-org/blob/cbad9b3b9c8b1531dabf8aeb792d7049b2eaf744/lib/rake/build.rake#L24

namely, js asset digesting (part of `build:dist`), rails asset digesting, and rails asset compilation. the constraint on the 2nd bullet above is because we can't safely digest rails assets without digesting js assets. So, if all you want is to `compile` rails assets, then it is possible we could separate `digest` from `compile` so that we can still build apps tests faster. or, perhaps you know a reason that `compile` and `digest` settings need to match each other, in which case never mind :-) you don't have to untangle this right now, but my point is that there could be opportunity to speed the build back up without breaking rails 6.1.

> I should also clarify that this adds 3.5 minutes specifically to the unit test container, which right now takes about 40 minutes less time to complete than the UI test container.

I agree that the added time isn't a big deal right now, but would hate for a win of this size to get lost as we strive to get our whole build time for both containers down to ~20-30 minutes, at which point 3.5 min would become relevant. so, I think it is worth documenting in some form in the comments near the lines you are removing.

Sorry for such a long response to a simple change! it's always harder when things are being taken away 🤷

File: docker/unit_tests.sh
Comment (davidsbailey) [inline docker/unit_tests.sh:36]: LGTM after adding a comment here explaining how we might be able to squeeze a few minutes out of the build time here.

**Review by davidsbailey — APPROVED** [review]

**Review by sureshc — APPROVED** [review]

**Comment by Hamms** [comment]
Thank you very much Dave for digging into this! You're right that these configurations are a bit entangled, and I think that entanglement led me to some incorrect conclusions. Namely, I mistakenly believed that we need to _precompile_ assets in order for them to be picked up by the asset pipeline; I hadn't realized that the `config.assets.compile` setting gives the webserver the ability to lazily compile content.

Now that I understand that, I think this PR is in fact entirely unnecessary! The errors that I thought were being caused by us not correctly compiling assets were in fact caused by us attempting to serve assets that actually don't exist. I'm working on fixing those issues with https://github.com/code-dot-org/code-dot-org/pull/47097 and https://github.com/code-dot-org/code-dot-org/pull/47172, at which point I believe our use of the Rails asset pipeline will be in a good state!

Closing this PR in favor of the others.

**Comment by davidsbailey** [comment]
Sounds great Elijah, I'm glad you were able to get to the bottom of it!

**Pull request closed without merging**
